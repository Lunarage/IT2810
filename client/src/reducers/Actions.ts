import HttpClient from "../modules/HttpClient";
import { SessionStorageWrapper, LocalStorageWrapper } from "../modules/Storage";

export const toggle_loggedIn = (boolVal: boolean) => {
    return {
        type: "TOGGLE_LOGGED_IN",
        payload: boolVal
    } as const
};

export const set_username = ( text: string) => {
    return {
        type: "SET_USERNAME",
        payload: text
    } as const
};

export const logout = (boolVal: boolean) => {
    return {
        type: "LOGOUT",
        payload: boolVal
    } as const 
};
