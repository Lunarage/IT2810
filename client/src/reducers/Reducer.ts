import {
    toggle_loggedIn,
    logout,
    set_username,
    setStoreSearchState,
} from "./Actions";
import { createStore, Store } from "redux";
import { Movie } from "../types/DatabaseTypes";

type Actions =
    | ReturnType<typeof toggle_loggedIn>
    | ReturnType<typeof logout>
    | ReturnType<typeof set_username>
    | ReturnType<typeof setStoreSearchState>;

// The type for holding search information and result
export type StoreSearchState = {
    inputSearch: string;
    inputOrderDir: "ASC" | "DESC";
    inputTitleType: string;
    pageActivePage: number;
    pageTotalPages: number;
    pageDisabled: boolean;
    searchStatus: "none" | "waiting" | "success" | "failure";
    searchErrorMessage: string | null;
    searchMovies: Movie[];
};

//Creates the types AppState
export type AppState = {
    loggedIn: boolean;
    userName?: string | null;
    searchState: StoreSearchState;
};

//Sets initial root state for the app
const initialAppState: AppState = {
    loggedIn: false,
    userName: null,
    searchState: {
        inputSearch: "",
        inputOrderDir: "ASC",
        inputTitleType: "",
        pageActivePage: 1,
        pageTotalPages: 2,
        pageDisabled: false,
        searchStatus: "none",
        searchErrorMessage: null,
        searchMovies: [],
    },
};
//Root reducer for the App, handels the states loggedIn/Loggedout and userName
export const rootReducer = (
    state: AppState = initialAppState,
    action: Actions
) => {
    switch (action.type) {
        case "TOGGLE_LOGGED_IN":
            return {
                ...state,
                loggedIn: action.payload,
            };

        case "SET_USERNAME":
            return {
                ...state,
                userName: action.payload,
            };

        case "LOGOUT":
            return {
                ...state,
                loggedIn: action.payload,
                userName: null,
            };
        case "SET_SEARCH_STATE":
            return {
                ...state,
                searchState: action.payload,
            };
        default:
            return state;
    }
};
//Configure store
function configureStore(): Store<AppState> {
    return createStore(
        rootReducer,
        (window as any).__REDUX_DEVTOOLS_EXTENSION__ &&
            (window as any).__REDUX_DEVTOOLS_EXTENSION__()
    );
}

//Create and exports store
export const store = configureStore();
