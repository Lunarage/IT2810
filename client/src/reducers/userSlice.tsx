import { createSlice } from "@reduxjs/toolkit";
import HttpClient from "../modules/HttpClient";
import { SessionStorageWrapper, LocalStorageWrapper } from "../modules/Storage";

// redux komponent som lager en "slice" av store hvor vi kan definere en reducer og actions som handler om en bruker logget inn eller ikke
const loginSlice = createSlice({
    name: "login/logout",
    initialState: { value: false, username: null },

    reducers: {
        logIn: (state, action) => {
            const baseURL = "http://it2810-22.idi.ntnu.no:3000";
            const client = new HttpClient(baseURL);
            const localStorage = new LocalStorageWrapper();
            localStorage.set("username", action.payload);
            client.createUser(action.payload);
            state.value = true;
            state.username = action.payload;
        },
        logOut: (state) => {
            const localStorage = new LocalStorageWrapper();
            localStorage.remove("username");
            state.value = false;
            state.username = null;
        },
    },
});

export const { logIn, logOut } = loginSlice.actions;

export default loginSlice.reducer;

/* These sites are used as src for this code: 
https://codesandbox.io/s/github/reduxjs/redux-essentials-counter-example/tree/master/?from-embed=&file=/src/features/counter/counterSlice.js:49-83
https://redux.js.org/tutorials/essentials/part-2-app-structure */
