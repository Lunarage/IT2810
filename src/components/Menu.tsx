import React, {Component} from "react";

class Menu extends Component<{}, {}> {
    render() {
        return (
            <div className={"menu"}>
                <a>[SÃ¸k]</a> <a>[Min side]</a>
            </div>
        )
    }

}


export default Menu