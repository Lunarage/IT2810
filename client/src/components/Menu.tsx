import React from "react";
import { useSelector, RootStateOrAny, useDispatch } from "react-redux";
import { logOut } from "../reducers/userSlice";

//Dette komponentet returnerer menyknappene øverst på siden, under banneret. den får inn props med logikken til knappene fra App
//Den tar også inn informasjon fra store og viser ulike knapper basert på dette
export const Menu = (props: Props) => {
  //Henter informasjon fra store om en bruker er logget inn og logut action fra userSlice
  const dispatch = useDispatch();
  const isLoggedIn = useSelector((state: RootStateOrAny) => state.loggedIn)
    .value;

  const onLogoutButtonClicked = () => {
    dispatch(logOut());
  };
  //if statement for om man skal vise en login eller loggut knapp basert på state
  let button;
  if (isLoggedIn) {
    button = (
      <button className="login-button" onClick={onLogoutButtonClicked}>
        Log out
      </button>
    );
  } else {
    button = (
      <button className="login-button" onClick={props.onLoginPageClick}>
        Log in
      </button>
    );
  }

  return (
    <div className={"menu"}>
      <a tabIndex={0} className="home-button" onClick={props.onHomePageClick}>
        [Home]
      </a>
      <a
        tabIndex={0}
        className="search-button"
        onClick={props.onSearchPageClick}
      >
        [Search]
      </a>
      {button}
      {isLoggedIn === true && (
        <a
          tabIndex={0}
          className="my-page-button"
          onClick={props.onMyPageClick}
        >
          [My page]
        </a>
      )}
    </div>
  );
};

export default Menu;

interface Props {
  onHomePageClick(): void;
  onSearchPageClick(): void;
  onMyPageClick(): void;
  onLoginPageClick(): void;
}

interface State {}
