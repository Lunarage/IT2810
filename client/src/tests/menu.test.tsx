import { render } from '@testing-library/react';
import Menu from "../components/Menu"
import React from 'react';
import { Provider } from "react-redux"
import { store } from "../reducers/Reducer"

test('Menu renders', () => {
    const {container} = render(
      <Provider store={store}><Menu/></Provider>);
    expect(container).toMatchSnapshot();
  });
