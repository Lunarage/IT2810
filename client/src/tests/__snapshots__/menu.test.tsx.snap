// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu snapshots Menu renders correctly when logged in 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="site-menu"
      >
        <p>
          Home
        </p>
        <a
          class="search-button"
          tabindex="0"
        >
          Search
        </a>
        <a
          class="my-page-button"
          tabindex="0"
        >
          My Page
        </a>
        <button
          class="login-button"
        >
          Log out
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="site-menu"
    >
      <p>
        Home
      </p>
      <a
        class="search-button"
        tabindex="0"
      >
        Search
      </a>
      <a
        class="my-page-button"
        tabindex="0"
      >
        My Page
      </a>
      <button
        class="login-button"
      >
        Log out
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Menu snapshots Menu renders correctly when not logged in 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="site-menu"
      >
        <p>
          Home
        </p>
        <a
          class="search-button"
          tabindex="0"
        >
          Search
        </a>
        <button
          class="login-button"
        >
          Log in
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="site-menu"
    >
      <p>
        Home
      </p>
      <a
        class="search-button"
        tabindex="0"
      >
        Search
      </a>
      <button
        class="login-button"
      >
        Log in
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
